ceno_zkvm/src/scheme/mock_prover.rs

```diff
@@ -1205,6 +1185,84 @@ Hints:
     }
 }
 
+fn compare_lkm<E, K>(
+    lkm_a: LkMultiplicityRaw<K>,
+    lkm_b: LkMultiplicityRaw<K>,
+    errors: &mut Vec<MockProverError<E, K>>,
+) where
+    E: ExtensionField,
+    K: LkMultiplicityKey,
+{
+    let lkm_a = lkm_a.into_finalize_result();
+    let lkm_b = lkm_b.into_finalize_result();
+
+    // Compare each LK Multiplicity.
+    for (rom_type, a_map, b_map) in izip!(ROMType::iter(), &lkm_a, &lkm_b) {
+        if *a_map != *b_map {
+            let a_keys: HashSet<_> = a_map.keys().collect();
+            let b_keys: HashSet<_> = b_map.keys().collect();
+
+            // Lookup missing in lkm_a.
+            b_keys.difference(&a_keys).for_each(|k| {
+                let b_count = b_map.get(k).unwrap();
+                if *b_count != 0 {
+                    errors.push(MockProverError::LkMultiplicityError {
+                        rom_type,
+                        key: **k,
+                        count: *b_count as isize,
+                    })
+                }
+            });
+
+            // Lookup missing in lkm_b.
+            a_keys.difference(&b_keys).for_each(|k| {
+                let a_count = a_map.get(k).unwrap();
+                if *a_count != 0 {
+                    errors.push(MockProverError::LkMultiplicityError {
+                        rom_type,
+                        key: **k,
+                        count: -(*a_count as isize),
+                    })
+                }
+            });
+
+            // Count of specific lookup differ lkm_b and lkm_a.
+            a_keys.intersection(&b_keys).for_each(|k| {
+                let a_count = a_map.get(k).unwrap();
+                let b_count = b_map.get(k).unwrap();
+
+                if a_count != b_count {
+                    errors.push(MockProverError::LkMultiplicityError {
+                        rom_type,
+                        key: **k,
+                        count: (*b_count as isize) - (*a_count as isize),
+                    })
+                }
+            });
+        }
+    }
+}
+
+fn print_errors<E: ExtensionField, K: LkMultiplicityKey>(
+    errors: &[MockProverError<E, K>],
+    wits_in: &[ArcMultilinearExtension<E>],
+    wits_in_name: &[String],
+    panic_on_error: bool,
+) {
+    println!("======================================================");
+    for (count, error) in errors.iter().dedup_with_count() {
+        error.print(wits_in, wits_in_name);
+        if count > 1 {
+            println!("Error: {} duplicates hidden.", count - 1);
+        }
+    }
+    println!("Error: {} constraints not satisfied", errors.len());
+    println!("======================================================");
+    if panic_on_error {
+        panic!("(Unexpected) Constraints not satisfied");
```

Comment:
If `panic_on_error` is set, we should probably put the full error message into the panic message?  (In that case, we don't need to println the error message, just put it in the panic only.)

Reply:
Sometimes the errors can be a lot and it can go beyond the terminal lines limit (if it's set low) and the following default rust panic stderr line would not be easily visible or lost.

```
thread 'main' panicked at /Users/.../mock_prover.rs:919:9:
```

That's why the panic message was purposefully kept short summary so that when program terminates, it can be obvious it was a panic and at which line, and dev can look at logs before to learn more details about what caused "(Unexpected) Constraints not satisfied".

But if it doesn't seem helpful, for better code we can put everything in panic.